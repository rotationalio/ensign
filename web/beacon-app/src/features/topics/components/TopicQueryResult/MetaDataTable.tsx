import { Trans } from '@lingui/macro';
import { Table } from '@rotational/beacon-core';
import { ErrorBoundary } from '@sentry/react';

import { getProjectQueryMetaData } from '../../utils';

type MetaDataTableProps = {
  metadataResult: any;
};

const MetaDataTable = ({ metadataResult }: MetaDataTableProps) => {
  const initialColumns: any = [
    {
      Header: '',
      accessor: 'key',
    },
    {
      Header: '',
      accessor: 'value',
    },
  ];
  return (
    <div className="mx-4">
      <ErrorBoundary
        fallback={
          <div className="item-center my-auto flex w-full text-center font-bold text-danger-500">
            <p>
              <Trans>
                Sorry we are having trouble listing the meta data for your event, please refresh the
                page and try again.
              </Trans>
            </p>
          </div>
        }
      >
        <div className="max-h-[150px]  overflow-x-hidden">
          <Table
            columns={initialColumns}
            data={getProjectQueryMetaData(metadataResult)}
            theadClassName="hidden"
            tdClassName="first:font-bold"
            data-cy="query-meta-data-table"
          />
        </div>
      </ErrorBoundary>
    </div>
  );
};

export default MetaDataTable;
