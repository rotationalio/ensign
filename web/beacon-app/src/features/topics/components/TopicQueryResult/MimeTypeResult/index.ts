export { default as BinaryResult } from './BinaryResult';
export { default as XMLResult } from './XMLResult';
