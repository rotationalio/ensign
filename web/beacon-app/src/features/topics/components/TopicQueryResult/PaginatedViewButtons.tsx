import { Trans } from '@lingui/macro';
import { Button } from '@rotational/beacon-core';
import { ImNext2, ImPrevious2 } from 'react-icons/im';

const PaginatedViewButtons = () => {
  return (
    <div className="mt-2 flex justify-between">
      <Button type="button">
        <div className="flex justify-center">
          <ImPrevious2 fontSize={20} />
          <Trans>Previous</Trans>
        </div>
      </Button>
      <Button type="button">
        <div className="flex justify-center">
          <Trans>Next</Trans>
          <ImNext2 fontSize={20} />
        </div>
      </Button>
    </div>
  );
};

export default PaginatedViewButtons;
