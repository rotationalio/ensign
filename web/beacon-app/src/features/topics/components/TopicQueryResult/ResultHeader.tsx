import { t, Trans } from '@lingui/macro';
import { mergeClassnames } from '@rotational/beacon-core';
import { FC } from 'react';
interface ResultHeaderProps {
  mimeType: string;
  eventType: string;
  isBase64Encoded: boolean;
}

const ResultHeader: FC<ResultHeaderProps> = ({ mimeType, eventType, isBase64Encoded }) => {
  const isParsable =
    isBase64Encoded && mimeType !== 'application/json' && mimeType !== 'application/xml';

  return (
    <div className="mx-auto flex h-12 flex-row items-center justify-between bg-[#2F4858]/70 p-4">
      <div className="flex flex-row">
        <p className="font-bold text-white">
          <Trans>MIME Type:</Trans>
        </p>
        <p
          className={mergeClassnames('ml-2', isParsable ? 'text-[#FCF77C]' : ' text-white')}
          data-cy="result-mime-type"
        >
          {isParsable ? t`Could not parse. Rendered as base-64 encoded data.` : mimeType ?? 'N/A'}
        </p>
      </div>
      <div className="flex flex-row">
        <p className="font-bold text-white">
          <Trans>Event Type & Version: </Trans>
        </p>
        <p className="ml-2 text-white" data-cy="result-event-type">
          {eventType ?? 'N/A'}
        </p>
      </div>
    </div>
  );
};

export default ResultHeader;
