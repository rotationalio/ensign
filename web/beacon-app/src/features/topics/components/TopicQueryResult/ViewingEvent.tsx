import { Trans } from '@lingui/macro';
import { Heading } from '@rotational/beacon-core';

import { CardListItem } from '@/components/common/CardListItem';

import MetaDataTable from './MetaDataTable';

interface MetaDataProps {
  totalResults: number;
  totalEvents: string;
}

const ViewingEvent = ({ totalResults, totalEvents }: MetaDataProps) => {
  return (
    <div className="mt-8">
      <Heading as="h2" className="mb-2 text-lg font-semibold">
        <Trans>
          Query Results
          <span className="ml-1 font-normal">
            {' '}
            - {String(totalResults ?? 'N/A')} of {totalEvents ?? 'N/A'}
          </span>
        </Trans>
      </Heading>
      <CardListItem className="!rounded-none">
        <p>
          <Trans>Viewing Event: 1 of 10</Trans>
        </p>
        <p className="pt-2 font-semibold">
          <Trans>Meta Data</Trans>
        </p>
        <MetaDataTable />
      </CardListItem>
    </div>
  );
};

export default ViewingEvent;
