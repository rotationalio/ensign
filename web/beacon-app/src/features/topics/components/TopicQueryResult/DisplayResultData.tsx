// this component is used to display the result data according to the mimetype

import React from 'react';

import { MIME_TYPE } from '@/constants';

import { BinaryResult } from './MimeTypeResult';
interface DisplayResultDataProps {
  result: any;
  mimeType: string;
  isBase64Encoded?: boolean;
}

// Default component that should be rendered if the mimetype is binary based
const renderDefaultResultComponent = (result: any) => {
  if (result instanceof ArrayBuffer) {
    return <BinaryResult data={result} />;
  }

  return <>{result}</>;
};

const DisplayResultData: React.FC<DisplayResultDataProps> = ({ result, mimeType }) => {
  switch (mimeType) {
    case MIME_TYPE.JSON:
      return <>{JSON.stringify(result, null, 2)}</>; // TODO: add syntax highlighting with  sc-19457
    case MIME_TYPE.XML:
      return <>{result}</>; // TODO: beautify xml with 19456
    default:
      return renderDefaultResultComponent(result);
  }
};

export default DisplayResultData;
