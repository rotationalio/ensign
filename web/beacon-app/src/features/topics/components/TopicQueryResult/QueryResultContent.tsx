import { t } from '@lingui/macro';
import React from 'react';

//import { getXMLFixture , createBinaryFixture} from '../../__mocks__';
import DisplayResultData from './DisplayResultData';
interface QueryResultContentProps {
  result: any;
  mimeType: string;
  error?: any;
}

const QueryResultContent: React.FC<QueryResultContentProps> = ({ result, mimeType, error }) => {
  console.log('[] result', result); // added this avoid eslint error
  console.log('[] mimetype', mimeType);
  // commented out the below two lines to test the binary result
  // const mockMimeType = 'application/octet-stream';
  // const mockResult = createBinaryFixture();
  // commented out the above two lines and uncomment the below two lines to test the default result
  // result = result ?? mockResult;
  // mimeType = mimeType ?? mockMimeType;
  // commented out the above two lines and uncomment the below two lines to test the XML result
  // result = result ?? getXMLFixture();
  // mimeType = mimeType ?? 'application/xml';

  return (
    <div className="shadow-md min-h-20 max-h-[480px] overflow-y-auto bg-black p-4 text-white">
      <pre className="font-base mx-auto">
        <code>
          {result && <DisplayResultData result={result} mimeType={mimeType} />}
          {error && t`No results found.`}
          {!result &&
            !mimeType &&
            t`No query result. Try the default query or enter your own query. See EnSQL documentation for example queries.`}
        </code>
      </pre>
    </div>
  );
};

export default QueryResultContent;
