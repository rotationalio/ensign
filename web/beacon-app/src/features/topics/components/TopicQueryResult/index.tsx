import usePaginateTopicQuery from '../../hooks/usePaginateTopicQuery';
import PaginatedViewButtons from './PaginatedViewButtons';
import QueryResultContent from './QueryResultContent';
import ResultHeader from './ResultHeader';
import ViewingEvent from './ViewingEvent';
interface TopicQueryResultProps {
  data: any;
  isFetching?: boolean;
  error?: any;
}

const TopicQueryResult = ({ data }: TopicQueryResultProps) => {
  const totalResults = data?.results?.length;

  const {
    result,
    isNextClickDisabled,
    isPrevClickDisabled,
    handleNextClick,
    handlePrevClick,
    counter,
  } = usePaginateTopicQuery(data);

  return (
    <div className="">
      <ViewingEvent
        totalResults={totalResults}
        totalEvents={data?.total_events}
        counter={counter}
      />
      <ResultHeader
        mimeType={result?.mimetype}
        eventType={result?.version}
        isBase64Encoded={result?.is_base64_encoded}
      />
      <QueryResultContent
        result={result?.data}
        mimeType={result?.mimetype}
        isBase64Encoded={result?.is_base64_encoded}
      />
      <PaginatedViewButtons
        onClickNext={handleNextClick}
        onClickPrevious={handlePrevClick}
        isNextDisabled={isNextClickDisabled}
        isPreviousDisabled={isPrevClickDisabled}
      />
    </div>
  );
};

export default TopicQueryResult;
