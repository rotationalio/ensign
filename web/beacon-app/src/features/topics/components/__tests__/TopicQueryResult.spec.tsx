/* eslint-disable testing-library/no-wait-for-multiple-assertions */
import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import React from 'react';
import { vi } from 'vitest';

import { getTopicQueryResponseMockData } from '../../__mocks__';
import TopicQueryResult from '../TopicQueryResult';

// mock Trans tag from lingui
vi.mock('@lingui/macro', () => ({
  t: (str: string) => str,
  Trans: ({ children }) => children,
}));

const renderComponent = () => {
  const propsMock = getTopicQueryResponseMockData();
  return render(<TopicQueryResult data={propsMock} />);
};

describe('TopicQueryResult', () => {
  it('should render the component with default value', () => {
    const { container } = renderComponent();
    expect(container).toMatchSnapshot();
  });

  it('should render with the first result data', async () => {
    renderComponent();

    expect(screen.getByText('10 results of 11 total')).toBeInTheDocument();
    expect(screen.getByText('Viewing Event: 1 of 10')).toBeInTheDocument();
    expect(screen.getByText('text/plain')).toBeInTheDocument();
    expect(screen.getByText('Message v1.0.0')).toBeInTheDocument();
    expect(screen.getByText('hello world')).toBeInTheDocument();

    expect(screen.getByTestId('prev-query-btn')).toBeDisabled();

    expect(screen.getByTestId('next-query-btn')).toBeEnabled();
  });

  it('should display the correct results when the next button is clicked', async () => {
    renderComponent();
    fireEvent.click(screen.getByTestId('next-query-btn')); // Simulate a click

    await waitFor(() => {
      expect(screen.getByText('text/csv')).toBeInTheDocument();
      expect(screen.getByText('Spreadsheet v1.1.0')).toBeInTheDocument();
      expect(screen.getByText('hello,world')).toBeInTheDocument();
      expect(screen.getByText('10 results of 11 total')).toBeInTheDocument();
      expect(screen.getByText('Viewing Event: 2 of 10')).toBeInTheDocument();
    });
    // Additional assertions after the state has loaded
    expect(screen.getByTestId('prev-query-btn')).toBeEnabled();
    expect(screen.getByTestId('next-query-btn')).toBeEnabled();
  });
});
