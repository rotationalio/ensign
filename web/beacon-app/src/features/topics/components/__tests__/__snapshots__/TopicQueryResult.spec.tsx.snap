// Vitest Snapshot v1

exports[`TopicQueryResult > should render the component with default value 1`] = `
<div>
  <div
    class=""
  >
    <div
      class="mt-8"
    >
      <h2
        class="mb-2 text-lg font-semibold"
      >
        Query Results -
        <span
          class="ml-1 font-normal"
          data-testid="query-result-count"
        >
           10 results of 11 total
        </span>
      </h2>
      <div
        class="sc-dkrFOg sc-hLBbgP  hdeBdD border-sm rounded-sm shadow-sm overflow-hidden bg-white padding-4 min-h-[200px] w-full border p-4 !rounded-none"
        data-testid="cardlistitem"
        style="border-radius: 4px;"
      >
        <div
          class="my-auto w-full rounded-[4px]"
        >
          <div
            class="sc-dkrFOg sc-jSUZER  cbRgnp"
          >
            <div
              class="space-y-3"
            >
              <p
                data-testid="view-event"
              >
                Viewing Event: 
                 1 of 10 
              </p>
              <p
                class="pt-2 font-semibold"
              >
                Meta Data
              </p>
              <div
                class="mx-4"
              >
                <div
                  class="sm:flex sm:gap-x-2"
                />
                <div
                  class="mt-4 flex flex-col"
                >
                  <div
                    class="-my-2 overflow-x-auto -mx-4 sm:-mx-6 lg:-mx-8"
                  >
                    <div
                      class="py-2 align-middle inline-block min-w-full sm:px-6 lg:px-8"
                    >
                      <div
                        class="shadow overflow-hidden border-b border-gray-200 sm:rounded-lg text-[12px]"
                      >
                        <table
                          class="min-w-full divide-y divide-gray-200"
                          role="table"
                        >
                          <thead
                            class="bg-white font-bold"
                          >
                            <tr
                              role="row"
                            >
                              <th
                                class="group px-6 py-3 text-left text-[12px] font-bold capitalize tracking-wider border-x-[1.5] border-gray-400/40"
                                colspan="1"
                                role="columnheader"
                                scope="col"
                                style="cursor: pointer;"
                                title="Toggle SortBy"
                              >
                                <div
                                  class="flex items-center justify-between text-[12px]"
                                >
                                  Key
                                  <span>
                                    <svg
                                      class="w-4 h-4 text-gray-400 opacity-0 group-hover:opacity-100"
                                      fill="currentColor"
                                      height="1em"
                                      stroke="currentColor"
                                      stroke-width="0"
                                      viewBox="0 0 320 512"
                                      width="1em"
                                      xmlns="http://www.w3.org/2000/svg"
                                    >
                                      <path
                                        d="M41 288h238c21.4 0 32.1 25.9 17 41L177 448c-9.4 9.4-24.6 9.4-33.9 0L24 329c-15.1-15.1-4.4-41 17-41zm255-105L177 64c-9.4-9.4-24.6-9.4-33.9 0L24 183c-15.1 15.1-4.4 41 17 41h238c21.4 0 32.1-25.9 17-41z"
                                      />
                                    </svg>
                                  </span>
                                </div>
                              </th>
                              <th
                                class="group px-6 py-3 text-left text-[12px] font-bold capitalize tracking-wider border-x-[1.5] border-gray-400/40"
                                colspan="1"
                                role="columnheader"
                                scope="col"
                                style="cursor: pointer;"
                                title="Toggle SortBy"
                              >
                                <div
                                  class="flex items-center justify-between text-[12px]"
                                >
                                  Value
                                  <span>
                                    <svg
                                      class="w-4 h-4 text-gray-400 opacity-0 group-hover:opacity-100"
                                      fill="currentColor"
                                      height="1em"
                                      stroke="currentColor"
                                      stroke-width="0"
                                      viewBox="0 0 320 512"
                                      width="1em"
                                      xmlns="http://www.w3.org/2000/svg"
                                    >
                                      <path
                                        d="M41 288h238c21.4 0 32.1 25.9 17 41L177 448c-9.4 9.4-24.6 9.4-33.9 0L24 329c-15.1-15.1-4.4-41 17-41zm255-105L177 64c-9.4-9.4-24.6-9.4-33.9 0L24 183c-15.1 15.1-4.4 41 17 41h238c21.4 0 32.1-25.9 17-41z"
                                      />
                                    </svg>
                                  </span>
                                </div>
                              </th>
                            </tr>
                          </thead>
                          <tbody
                            class="bg-white divide-y divide-gray-[#1C1C1C]"
                            role="rowgroup"
                          >
                            <tr
                              class="text-center items-center"
                            >
                              <td
                                class="px-auto py-4 whitespace-nowrap text-center h-[100px] items-center"
                                colspan="2"
                              >
                                <div
                                  class="flex text-center justify-center"
                                >
                                  <svg
                                    class="w-16 h-16 text-gray-500"
                                    fill="currentColor"
                                    height="1em"
                                    stroke="currentColor"
                                    stroke-width="0"
                                    viewBox="0 0 512 512"
                                    width="1em"
                                    xmlns="http://www.w3.org/2000/svg"
                                  >
                                    <path
                                      d="M256 8C119.033 8 8 119.033 8 256s111.033 248 248 248 248-111.033 248-248S392.967 8 256 8zm0 464c-110.28 0-200-89.72-200-200S145.72 56 256 56s200 89.72 200 200-89.72 200-200 200z"
                                    />
                                    <path
                                      d="M256 128c-44.183 0-80 35.817-80 80s35.817 80 80 80 80-35.817 80-80-35.817-80-80-80zm0 128c-26.51 0-48-21.49-48-48s21.49-48 48-48 48 21.49 48 48-21.49 48-48 48z"
                                    />
                                    <path
                                      d="M256 320c-8.837 0-16-7.163-16-16v-96c0-8.837 7.163-16 16-16s16 7.163 16 16v96c0 8.837-7.163 16-16 16z"
                                    />
                                  </svg>
                                </div>
                                <p
                                  class="mt-2 text-sm text-gray-500"
                                >
                                  No data available
                                </p>
                              </td>
                            </tr>
                          </tbody>
                        </table>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="mx-auto flex h-12 flex-row items-center justify-between bg-[#2F4858]/70 p-4"
    >
      <div
        class="flex flex-row"
      >
        <p
          class="font-bold text-white"
        >
          MIME Type:
        </p>
        <p
          class="ml-2 text-white"
        >
          text/plain
        </p>
      </div>
      <div
        class="flex flex-row"
      >
        <p
          class="font-bold text-white"
        >
          Event Type & Version: 
        </p>
        <p
          class="ml-2 text-white"
        >
          Message v1.0.0
        </p>
      </div>
    </div>
    <div
      class="shadow-md min-h-20 max-h-[480px] overflow-y-auto bg-black p-4 text-white"
    >
      <pre
        class="font-base mx-auto"
      >
        <code>
          hello world
        </code>
      </pre>
    </div>
    <div
      class="mt-2 flex justify-between"
    >
      <button
        class="sc-bcXHqe jLAqbK line-height-1.75 cursor-pointer text-white font-size-14 px-4 py-2 min-h-[28px] bg-inherit transition-colors duration-200 ease-in-out rounded-5 focus:outline-none text-[14px] font-bold"
        data-testid="prev-query-btn"
        disabled=""
        type="button"
      >
        <div
          class="flex justify-center"
        >
          <svg
            fill="currentColor"
            font-size="20"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            version="1.1"
            viewBox="0 0 16 16"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M4 14v-12h2v5.5l5-5v11l-5-5v5.5z"
            />
          </svg>
          Previous
        </div>
      </button>
      <button
        class="sc-bcXHqe iZHJaz line-height-1.75 cursor-pointer text-white font-size-14 px-4 py-2 min-h-[28px] bg-inherit transition-colors duration-200 ease-in-out rounded-5 focus:outline-none text-[14px] font-bold"
        data-testid="next-query-btn"
        type="button"
      >
        <div
          class="flex justify-center"
        >
          Next
          <svg
            fill="currentColor"
            font-size="20"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            version="1.1"
            viewBox="0 0 16 16"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M12 2v12h-2v-5.5l-5 5v-11l5 5v-5.5z"
            />
          </svg>
        </div>
      </button>
    </div>
  </div>
</div>
`;
